buffer/src/main/java/io/netty/buffer/api/Buffer.java

```diff
@@ -0,0 +1,688 @@
+     * <p>
+     * This method does not read or modify the {@linkplain #writerOffset() write offset} or the
+     * {@linkplain #readerOffset() read offset}, nor is the position of the destination buffer changed.
+     * <p>
+     * The position and limit of the destination byte buffer are also ignored, and do not influence {@code destPos}
+     * or {@code length}.
+     *
+     * @param srcPos The byte offset into this buffer from where the copying should start; the byte at this offset in
+     *              this buffer will be copied to the {@code destPos} index in the {@code dest} array.
+     * @param dest The destination byte buffer.
+     * @param destPos The index into the {@code dest} array from where the copying should start.
+     * @param length The number of bytes to copy.
+     * @throws NullPointerException if the destination array is null.
+     * @throws IndexOutOfBoundsException if the source or destination positions, or the length, are negative,
+     * or if the resulting end positions reaches beyond the end of either this buffer, or the destination array.
+     */
+    void copyInto(int srcPos, ByteBuffer dest, int destPos, int length);
+
+    /**
+     * Copies the given length of data from this buffer into the given destination buffer, beginning at the given
+     * source position in this buffer, and the given destination position in the destination buffer.
+     * <p>
+     * This method does not read or modify the {@linkplain #writerOffset() write offset} or the
+     * {@linkplain #readerOffset() read offset} on this buffer, nor on the destination buffer.
+     * <p>
+     * The read and write offsets of the destination buffer are also ignored, and do not influence {@code destPos}
+     * or {@code length}.
+     *
+     * @param srcPos The byte offset into this buffer from where the copying should start; the byte at this offset in
+     *              this buffer will be copied to the {@code destPos} index in the {@code dest} array.
+     * @param dest The destination buffer.
+     * @param destPos The index into the {@code dest} array from where the copying should start.
+     * @param length The number of bytes to copy.
+     * @throws NullPointerException if the destination array is null.
+     * @throws IndexOutOfBoundsException if the source or destination positions, or the length, are negative,
+     * or if the resulting end positions reaches beyond the end of either this buffer, or the destination array.
+     */
+    void copyInto(int srcPos, Buffer dest, int destPos, int length);
+
+    /**
+     * Writes into this buffer, all the readable bytes from the given buffer.
+     * This updates the {@linkplain #writerOffset() write offset} of this buffer, and the
+     * {@linkplain #readerOffset() reader offset} of the given buffer.
+     *
+     * @param source The buffer to read from.
+     * @return This buffer.
+     */
+    default Buffer writeBytes(Buffer source) {
```

Comment:
nit: Why does this method have bytes in the name, while the complement methods above call it copyInto?   Maybe this might be `copyFrom`?  Or `copyBytesFrom` to indicate the indexes are not modified?

Reply:
The read*/write* family of methods modify the offsets, while the copy family does not.

Reply:
One idea you might consider is the "at" naming method.   It would make the BufferAccessors below also have a more consistent naming scheme (which is using "get", but Buffer is using "copy").    For example: 

* writeBytes(Buffer)
* writeBytesAt(pos, Buffer, len)
* readByte()
* readByteAt(pos)

The At variants would be a predictable, easy to remember name.    I don't think most people would assume the offsets would change when using the At variant.  I haven't tried it out with all of the methods but please consider it.

Reply:
I haven't heard of confusion about the get/set vs. read/write naming conventions from anyone else, so I'll keep it as is for now.